docs/user-guide/app_dependencies.md

```diff
@@ -0,0 +1,121 @@
+# Application dependencies
+
+Starting from Argo CD v2.9, you can natively model dependencies between Applications. This new mechanism is intended to replace the prior way of modeling dependencies when using the app-of-apps pattern with sync waves on the Application custom resources.
+
+Using the new dependency mechanism, users are not bound to the app-of-apps pattern anymore and can define dependencies between applications using application selectors to define a directed, acyclic graph for dependency modeling. This means that an Application which defines one or more dependencies do not only depend on its direct dependencies, but to the dependencies of its dependencies as well.
+
+Application dependencies support both, manual and automatic sync policies.
+
+## Feature state and caveats
+
+This feature is in [beta state](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md#beta). It will be polished across the upcoming releases.
+
+Bug reports and enhancement requests are most welcome.
+
+## Limitations and known issues
+
+A list of things that are good to know before getting started with Application dependencies:
+
+* An Application can only depend on other Applications that use the same AppProject and are in the same namespace as the dependent Application.
+* A circular dependency defined anywhere in the dependency chain is considered to be an error and will prevent any application in the chain from syncing.
+* Syncing an application that is dependent on other apps will not trigger a sync of the dependencies. However, manual sync of a dependent application whose dependencies are not satisfied will block.
+* Dependencies currently can only be configured in a declarative manner; the UI and CLI do not yet have first class support for configuring dependencies.
+
```

Comment:
Is there a way to also extend this so that `Application`s can be deleted in reverse order? IMO it's not super interesting to be able to depend on another application, but on the other hand it becomes super important to delete applications in reverse order (e.g for the example above, if `aws-load-balancer-controller` gets deleted before the dependent apps, we'd have orphaned load balancers).

Ideally there'd be a way to set the dependency policy, something like

1. `both` (use the DAG both when syncing an application and deleting it)
2. `delete-only` 
3. `sync-only`

Reply:
IMO, we should block a deletion of an App in the first iteration itself if there are any dependent applications on the current application. 

The deletion of an Application in the dependency chain could lead to sync issues in the applications depending on the deleted application. For example, consider the dependency chain, A -> B -> C -> D (A depends on B, B depends on C, etc.). If the application B is deleted and the `blockOnEmpty` is not set to true, this will break the Application A getting synced.

WDYT?

Reply:
While I definitely see the issue behind of what you're saying, I'm not sure I agree with the criticality. 

I thought a little bit more about the deletion subject recently, and I think we might not want do bring it as part of the MVP. My concerns basically revolve around a couple of thoughts, that potentially need further discussions, investigations and validations:

* Deletion in Kubernetes can only be blocked by setting a finalizer on a resource or by modeling owner references with `blockOwnerDeletion` set to true. Owner references only work within the same namespace, so I think they're out of the picture for now. Leaves us with finalizers, and as long as a finalizer is set, the Kubernetes GC will not reap a given resource. So we'd need to put a new finalizer on every Application that is a dependency of another Application. However, due to the dynamic nature of this finalizer, it's not clear what that means in term of sync status for the Application (for example, user has put a resources finalizer manually to the application, and we inject a dependency finalizer).

* Once a deletion is triggered, it cannot be easily aborted (if at all). For example, if a user deletes an Application that is a dependency of something else, the Application will stay in terminating mode as long as the dependent still exist. The only way to forcefully delete that Application is to remove the finalizer from the resource, but right now, there is no way I'm aware of to do this via the UI or CLI. That might come unexpected for many, especially when this finalizer is introduced to the Application on behalf of another Application (the dependent), and not in direct control of the Application's owner.

Not saying this is impossible or we shouldn't do it, but IMO it's out of scope for the MVP. The problem we are  trying to solve first is to allow ordered sync of Applications without using the app-of-apps pattern and sync waves.

The deletion topic makes a good follow up enhancement from my point of view.

Reply:
It makes sense to take care of it in the next iteration. But in that case, lets explicitly call it out in the docs that the deletion of any application from the dependency chain can break the application sync and the users need to be careful before deleting the application.

Reply:
I called it out in the docs.